### PR TITLE
gitlab: build sphinx doc in separate job

### PR DESCRIPTION
**Pull request by SkySkimmer**
The system to avoid passing BOOT to coqtop while sphinxing is a bit inelegant, suggestions welcome.

### CODE REVIEW

**Comment by ejgallego** [comment]
Umm, I was under the impression that we didn't want to do this as to get a full package in phase 1, right?

I am OK with any choice.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -48,7 +49,7 @@ before_script:
 
     - echo 'start:coq.build'
     - make -j "$NJOBS" byte
-    - make -j "$NJOBS"
+    - make -j "$NJOBS" world $EXTRA_TARGET
```
Comment (ejgallego) [inline .gitlab-ci.yml]: Psss, I am not happy to see such hacks, in fact, `coq_config.py` should have never been introduced in the first place like it is now [a configure target] but instead it should have been generated using a make rule [as the only thing it contains is the version]. But when I saw it I didn't want to complain as they doc people had enough of a party back then.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -37,6 +37,7 @@ before_script:
     paths:
       - _install_ci
       - config/Makefile
+      - config/coq_config.py
```
Comment (ejgallego) [inline .gitlab-ci.yml]: Psss, I am not happy to see such hacks, in fact, coq_config.py should have never been introduced in the first place like it is now [a configure target] but instead it should have been generated using a make rule [as the only thing it contains is the version]. But when I saw it I didn't want to complain as they doc people had enough of a party back then.

File: Makefile.common [hunk 3]
```diff
@@ -14,6 +14,9 @@
 # Executables
 ###########################################################################
 
+COQBIN:=$(PWD)/bin/
+export COQBIN
+
```
Comment (ejgallego) [inline Makefile.common]: This looks very suspicious, in particular for non-local builds.

**Comment by SkySkimmer** [comment]
I thought it was more about not having to figure out the separate 
building while migrating.

Gaëtan Gilbert

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -155,7 +168,9 @@ before_script:
 build:base:
   <<: *build-template
   variables:
-    COQ_EXTRA_CONF: "-native-compiler yes -coqide opt -with-doc yes"
+    COQ_EXTRA_CONF: "-native-compiler yes -coqide opt"
+    # coqdoc for stdlib, until we know how to build it from installed Coq
+    EXTRA_TARGET: "stdlib"
```
Comment (ejgallego) [inline .gitlab-ci.yml]: Why is this new target needed?

**Comment by SkySkimmer** [comment]
To build it without building sphinx

Gaëtan Gilbert

**Review by ejgallego — COMMENTED** [review]
I think this PR will have a better time if you actually split it in 2 [one for the changes to Make / Sphinx] and other for the changes to CI.

**Comment by ejgallego** [comment]
> I thought it was more about not having to figure out the separate building while migrating.

I have no idea, all that I have is vague recollection. So CI-wise, my only question is what about the artifacts and here @Zimmi48 is the main person linking at them.

Am I correct in the current status of this PR won't persist the documentation artifact?

**Comment by SkySkimmer** [comment]
If people are using the artifacts we could make the doc job artifact the 
whole install directory. It's a bit of redundant upload and if the main 
use of artifacts is looking at the doc it makes the download slower but 
not in a huge way.
I don't think anyone is actually using the artifacts though, considering 
they have some silly path baked in.

My main goal with this PR is to avoid having sphinx issues like the 
timeout one stop the entire pipeline.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> My main goal with this PR is to avoid having sphinx issues like the
> timeout one stop the entire pipeline.

Indeed seems like a sensible goal, and I had also thought of that, not because of the errors but also due to the extra latency Sphinx introduced.

Thus, in general terms I'm OK with this PR provided the artifact issue is clarified [cc:  #7161 ]

**Comment by ejgallego** [comment]
As far as my competences go, I can take care of the patch except for the `COQBIN` change, for that, you either revert it or get a `Makefile` maintainer to approve it.

**Comment by Zimmi48** [comment]
I would also like to see the patch split in two: one part fixing #6505 and one part using this for GitLab CI.
I want to keep access to the doc artifacts (especially the manual). I think they are very useful to be able to review a documentation PR.
Overall, I'm in favor of this PR.

File: .gitlab-ci.yml (hunk 4 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: Ping.

File: Makefile.common (hunk 3 above)
Comment (ejgallego) [reply Makefile.common]: Is this still needed?

File: .gitlab-ci.yml (hunk 4 above)
Comment (SkySkimmer) [reply .gitlab-ci.yml]: >To build it without building sphinx

**Comment by ejgallego** [comment]
I think @Zimmi48 's concerns about failing are a bit less; this looks OK to me, however I would like to understand a bit more what's going on, I am not sure I follow the targets/path setting here.

Could you answer my questions @SkySkimmer ? Merci.

File: Makefile.common (hunk 3 above)
Comment (SkySkimmer) [reply Makefile.common]: It's so we can override it in gitlab: `make COQBIN=...`

Comment (ejgallego) [reply Makefile.common]: But my worry here is that the path may not be correct, right? In non-local builds we should use `DESTDIR`?

**Comment by ejgallego** [comment]
I am unable to understand this, I guess I am too tired to look at Makefiles. I'll try again soon, but if someone can review the make part I'll merge happily.

File: Makefile.common (hunk 3 above)
Comment (gares) [reply Makefile.common]: I'm against this change. If you need `$COQBIN` to hold a good value, export it in the CI script.

File: Makefile.doc [hunk 5]
```diff
@@ -55,7 +55,7 @@ endif
 
 sphinx: $(SPHINX_DEPS)
 	$(SHOW)'SPHINXBUILD doc/sphinx'
-	$(HIDE)COQBIN="$(PWD)/bin" $(SPHINXBUILD) -W -b html $(ALLSPHINXOPTS) doc/sphinx $(SPHINXBUILDDIR)/html
```
[the remark this answers is not in the document]
Comment (gares) [reply Makefile.doc]: Oh, I did not know COQBIN was used internally, now I better understand what you are trying to do.
I'll be more explicit in the review

**Review by gares — CHANGES_REQUESTED** [review]
Put `COQBIN?=$(shell pwd)/bin` somewhere, do not export it and invoke sphinx as `COQBIN=$(COQBIN) $(SPHIX...)`

AFAIK COQBIN is not used internally in the makefile, it is just supported by coq_makefile. I guess this is why Clement knew about this variable and used it here. I dont think he should have.

The better solution would be to drop COQBIN, use `PATH` and have the makefile add to PATH `$(PWD)/bin` before running sphinx. Also, is seems the python class accepts an explicit path to coqtop, but I don'k know if there is a command line flag to sphinx that set it (in that case you don't even need to set PATH).

**Comment by SkySkimmer** [comment]
New version.

File: Makefile.common [hunk 6]
```diff
@@ -14,6 +14,8 @@
 # Executables
 ###########################################################################
 
+COQBIN:=$(CURDIR)/bin/
```
Comment (gares) [inline Makefile.common]: I prefer `?=`. For example:

```
gares@gargamel:~$ cd /tmp/
gares@gargamel:/tmp$ cat > Makefile
X:=3
all:
	echo $(X)
gares@gargamel:/tmp$ make
echo 3
3
gares@gargamel:/tmp$ make X=4
echo 4
4
gares@gargamel:/tmp$ X=4 make
echo 3
3
gares@gargamel:/tmp$ cat > Makefile
X?=3
all:
	echo $(X)
gares@gargamel:/tmp$ make
echo 3
3
gares@gargamel:/tmp$ make X=4
echo 4
4
gares@gargamel:/tmp$ X=4 make
echo 4
4
gares@gargamel:/tmp$ 
```

Comment (SkySkimmer) [reply Makefile.common]: I prefer `:=` so that I don't have to worry about unclean environment.

File: Makefile.common (hunk 3 above)
Comment (ejgallego) [reply Makefile.common]: I agree with Enrico, this looks fishy somehow.

**Comment by SkySkimmer** [comment]
It was changed.

Gaëtan Gilbert

**Review by ejgallego — CHANGES_REQUESTED** [review]
I understand this better now, indeed `COQBIN` should not be used in core Makefiles, as this will break for example building the documentation with Dune, which takes care to properly bring the right coqtop into scope.

I think sphinx should rely on PATH, or understand the value provided in `COQTOPBEST` which is the proper variable AFAICS.

**Comment by SkySkimmer** [comment]
I don't understand what you want done. It sounds like a refactoring that's out of scope for this PR.
I just want the sphinx errors to stop blocking my builds.

**Comment by ejgallego** [comment]
> I don't understand what you want done. It sounds like a refactoring that's out of scope for this PR.
> I just want the sphinx errors to stop blocking my builds.

Sorry if our two request for changes are not clear. I think that the modification we are asking for is:
- don't use `COQBIN`, it was not meant to be used by the build system and its introduction is an error;
- so drop the COQBIN part, and instead either use the already existing `COQTOPBEST` or set the path properly and let Sphinx use the `coqtop` in-scope.

**Comment by SkySkimmer** [comment]
Updated

File: Makefile.common [hunk 7]
```diff
@@ -10,6 +10,9 @@
 
 -include config/Makefile
 
+PATH:=$(CURDIR)/bin:$(PATH)
+export PATH
+
```
Comment (ejgallego) [inline Makefile.common]: Sorry if I am slow here, why do we need to update the `PATH` here?

[Note that I am not saying this is a bad thing to do].

**Review by ejgallego — DISMISSED** [review]
Looks better now, I still have one question tho.

**Comment by ejgallego** [comment]
I didn't mean to have a "request changes" here, only a commnet, but Github did like that...

File: Makefile.common (hunk 7 above)
Comment (SkySkimmer) [reply Makefile.common]: We need to update it somewhere, here seems like the right place.

Comment (ejgallego) [reply Makefile.common]: What component does need the new path? AFAICS we also set `PATH` in the Sphinx call, so this should not be needed, right?

I am fine with setting PATH globally and in fact this is how the Dune system works, but then we'd need to tweak quite a few items in the makefile. IMHO not worth it.

**Review by gares — APPROVED** [review]

**Comment by ejgallego** [comment]
I repeat here as I don't see my question:

Why do we need to export `PATH`, what component does need it? AFAICS we also set `PATH+  in the Sphinx call, so the export should not be needed, right?

I am fine with setting `PATH` globally and in fact this is how the Dune system works, but then we'd need to tweak quite a few items in the makefile. IMHO not worth it.

**Comment by ejgallego** [comment]
Oh so certainly I looked at wrong diff or my memory failed today.

But this doesn't seem broken, right? In fact, I may be happy to merge it, but I'd like to see what @gares and @letouzey think of settting up PATH in makefile. Indeed, that interacts with many components and has the potential to simplify quite a few things.

**Comment by ejgallego** [comment]
> has the potential to simplify quite a few things.

On the other hand it is hard to predict what the interactions will be.

**Comment by SkySkimmer** [comment]
In the log for the new job it says things like "/usr/bin/x86_64-linux-gnu-ld: cannot open output file bin/ocamllibdep: No such file or directory"
A previous CI run failed but maybe it was spurious and led to me noticing the above.

**Comment by ejgallego** [comment]
> In the log for the new job it says things like "/usr/bin/x86_64-linux-gnu-ld: cannot open output file bin/ocamllibdep: No such file or directory"
> A previous CI run failed but maybe it was spurious and led to me noticing the above.

Ok let's have a look, indeed setting `PATH` just for Sphinx may be the way to go for the moment.

**Comment by ejgallego** [comment]
We need to improve more but this looks good to me. Note that this PR conflicts with #7684 , I prefer the naming there as IMHO we should prefix variables.

I propose we merge this and I rebase on top and tweak the variable.

**Review by ejgallego — APPROVED** [review]

**Pull request merged**